gnomad_constraint/utils/constraint.py

```diff
@@ -734,6 +734,52 @@ def calculate_mu_by_downsampling(
     return annotate_mutation_type(ht)
 
 
+def add_oe_lof_upper_rank_and_bin(
+    ht: hl.Table, use_mane_select_instead_of_canonical: bool = False
+) -> hl.Table:
+    """
+    Compute the rank, decile, and sextile of the lof oe upper confidence interval for MANE Select or canonical ensembl transcripts.
+
+    :param ht: Input Table with the value for the lof oe upper confidence interval stored in ht.lof.oe_ci.upper.
+    :param use_mane_select_instead_of_canonical: Use MANE Select rather than canonical transcripts for filtering the Table.
+    :return: Table with anntotations added for 'upper_rank', 'upper_bin_decile', and 'upper_bin_sextile'.
+    """
+    if use_mane_select_instead_of_canonical:
+        transcript_filter = ht.mane_select
+    else:
+        transcript_filter = ht.canonical
+
+    # Filter to only ensembl transcripts of the specified transcript_filter.
+    ms_ht = ht.filter((transcript_filter) & (ht.transcript.startswith("ENST")))
+
+    # Rank lof.oe_ci.upper in ascending order.
+    ms_ht = ms_ht.order_by(ms_ht.lof.oe_ci.upper).add_index(name="upper_rank")
+
+    # Determine decile and sextile bins.
+    n_transcripts = ms_ht.count()
+    ms_ht = ms_ht.annotate(
+        upper_bin_decile=hl.int(ms_ht.upper_rank * 10 / n_transcripts)
+    )
+    ms_ht = ms_ht.annotate(
+        upper_bin_sextile=hl.int(ms_ht.upper_rank * 6 / n_transcripts)
+    )
+
+    # Add rank and bin annotations back to original Table.
+    ms_ht = ms_ht.key_by(*list(ht.key))
```

Comment:
does ordering by `lof.oe_ci.upper` shuffle the keys? if it doesn't, you may be able to use this here
```
ms_ht =  ms_ht._key_by_assert_sorted(*list(ht.key))
```
to explicitly avoid another shuffle